include/zephyr/sys/zync.h

```diff
@@ -0,0 +1,386 @@
+	_wait_q_t waiters;
+	IF_ENABLED(Z_ZYNC_OWNER, (struct k_thread *owner;))
+	IF_ENABLED(CONFIG_POLL, (sys_dlist_t poll_events;))
+	struct k_spinlock lock;
+	struct k_zync_cfg cfg;
+	IF_ENABLED(CONFIG_ZYNC_RECURSIVE, (uint32_t rec_count;))
+	IF_ENABLED(Z_ZYNC_INTERNAL_ATOM, (k_zync_atom_t atom;))
+	IF_ENABLED(CONFIG_ZYNC_PRIO_BOOST, (int8_t orig_prio;))
+	IF_ENABLED(CONFIG_POLL, (bool pollable;))
+};
+
+#define Z_ZYNC_MVCLAMP(v) ((v) == 0 ? K_ZYNC_ATOM_VAL_MAX \
+			      : CLAMP((v), 0, K_ZYNC_ATOM_VAL_MAX))
+
+#define K_ZYNC_INITIALIZER(init, isfair, rec, prioboost, maxval) {	\
+	.cfg.atom_init = init,						\
+	IF_ENABLED(CONFIG_ZYNC_MAX_VAL,					\
+		   (.cfg.max_val = Z_ZYNC_MVCLAMP(maxval),))		\
+	IF_ENABLED(CONFIG_ZYNC_RECURSIVE,				\
+		   (.cfg.recursive = rec,))				\
+	IF_ENABLED(CONFIG_ZYNC_PRIO_BOOST,				\
+		   (.cfg.prio_boost = prioboost,))			\
+	IF_ENABLED(CONFIG_POLL,						\
+		   (.pollable = (init != 0),))				\
+	.cfg.fair = (isfair) }
+
+#define K_ZYNC_DEFINE(name, init, isfair, rec, prioboost, maxval)	\
+	STRUCT_SECTION_ITERABLE(k_zync, name) =				\
+		K_ZYNC_INITIALIZER(init, isfair, rec, prioboost, maxval);
+
+/** @brief Atomically modify a k_zync_atom
+ *
+ * This macro heads a code block which is responsible for assigning
+ * fields of "new_atom" in terms of the (repeatedly-re-read)
+ * "old_atom", attempting to set the value with `atomic_cas()`, and
+ * repeating the process if the value was changed from another
+ * context.  Code can exit the loop without setting any value by using
+ * a break statement.
+ *
+ * No other modification of the values inside a k_zync_atom from
+ * potentially simultaneous contexts is permitted.
+ *
+ * @note The atom_ptr argument is expanded multiple times in the macro
+ * body and must not have side effects.
+ *
+ * @param atom_ptr Pointer to a k_zync_atom to modify
+ */
+#define K_ZYNC_ATOM_SET(atom) 							\
+for (k_zync_atom_t old_atom = { .atomic = atomic_get(&(atom)->atomic) },	\
+ 		   new_atom = old_atom, done = {};				\
+     !done.atomic;								\
+     done.atomic = atomic_cas(&(atom)->atomic, old_atom.atomic, new_atom.atomic)\
+     , old_atom.atomic = done.atomic ?						\
+			 old_atom.atomic : atomic_get(&(atom)->atomic) 		\
+     , new_atom = done.atomic ? new_atom : old_atom)
+
+/** @brief Try a zync atom modification
+ *
+ * Attempts an atomic mod operation on the value field of a zync atom,
+ * as specified for k_zync() (but without clamping to a max_val other
+ * than the static field maximum).  Returns true if the modifcation
+ * was be made completely without saturation and if no other threads
+ * are waiting.  Will not otherwise modify the atom state, and no
+ * intermediate states will be visible to other zync code.
+ *
+ * @param atom A pointer to a zync atom to be modified
+ * @param mod A count to add to the atom value
+ * @return True if the atom was successfully modified, otherwise false
+ */
+static inline bool k_zync_try_mod(k_zync_atom_t *atom, int32_t mod)
+{
```

Comment:
General question: Given the fundamental nature of zync in this new model, is it worthwhile to attempt to leverage the "unlikely" and "likely" macros for the best case scenarios?

If so, I can understand that being a separate commit set.